checker/tests/resourceleak/MultipleMethodParamsMustCallAliasTest.java

```diff
@@ -0,0 +1,62 @@
+import java.io.*;
+import java.net.*;
+import org.checkerframework.checker.calledmethods.qual.*;
+import org.checkerframework.checker.mustcall.qual.*;
+import org.checkerframework.common.returnsreceiver.qual.*;
+
+class MultipleMethodParamsMustCallAliasTest {
+
+  void testMultiMethodParamsCorrect(@Owning InputStream in1, @Owning InputStream in2)
+      throws IOException {
+
+    ReplicaInputStreams r = new ReplicaInputStreams(in1, in2);
+
+    r.close();
+  }
+
+  // :: error: required.method.not.called
+  void testMultiMethodParamsWrong1(@Owning InputStream in1, @Owning InputStream in2)
+      throws IOException {
+
+    ReplicaInputStreams r = new ReplicaInputStreams(in1, in2);
+
+    in1.close();
+  }
+
+  // :: error: required.method.not.called
+  void testMultiMethodParamsWrong2(@Owning InputStream in1, @Owning InputStream in2)
+      throws IOException {
+
+    ReplicaInputStreams r = new ReplicaInputStreams(in1, in2);
+
+    in2.close();
+  }
+
+  // :: error: required.method.not.called
+  void testMultiMethodParamsWrong3(@Owning InputStream in1) throws IOException {
+    // :: error: required.method.not.called
+    Socket socket = new Socket("address", 12);
+    ReplicaInputStreams r = new ReplicaInputStreams(in1, socket.getInputStream());
+  }
+
+  class ReplicaInputStreams implements Closeable {
+
+    private final @Owning InputStream in1;
+    private final @Owning InputStream in2;
+
+    public @MustCallAlias ReplicaInputStreams(
+        @MustCallAlias InputStream i1, @MustCallAlias InputStream i2) {
+      this.in1 = i1;
+      this.in2 = i2;
+    }
+
+    @Override
+    @EnsuresCalledMethods(
```

Comment:
This is interesting.  I think this `@EnsuresCalledMethods` annotation gets verified because if the normal exit of `close()` is reached, then both `in1.close()` and `in2.close()` will be invoked.  But, if `in1.close()` throws an `IOException`, then `in2.close()` never gets called and `in2` might leak!  @kelloggm is this a possible soundness hole in the type system?

Reply:
This might be a soundness bug. `@EnsuresCalledMethods` is defined to mean that the given methods are called on the given expressions *if the method terminates successfully* (emphasis mine). That is, an `@EnsuresCalledMethods` annotation doesn't make any guarantees about exceptional exits.

However, we may have relied on a stronger contract when we decided how to handle owning fields. In particular, I think we use `@EnsuresCalledMethods` to mean "the given methods are either called or are the cause of any exceptional exit" - because close() can throw an IOException, we need that last clause to verify even a simple method that attempts to close a single resource and allows any exceptions to propagate. (A different question, in light of this test case, is whether we ought to verify that at all...)

Reply:
We'll fix this separately, #4838